lib/mix/tasks/compile.thrift.ex

```diff
@@ -68,21 +69,29 @@ defmodule Mix.Tasks.Compile.Thrift do
       |> Keyword.take([:include_paths, :namespace])
       |> Keyword.put_new(:namespace, "Thrift.Generated")
 
-    mappings =
-      input_files
-      |> Enum.map(&parse(&1, parser_opts))
-      |> Enum.reject(&is_nil/1)
-      |> extract_targets(output_path, opts[:force])
+    {_, diagnostics} =
+      result =
+      case parse(input_files, parser_opts) do
+        {[], []} ->
+          {:noop, []}
 
-    generate(manifest(), mappings, output_path, opts)
+        {groups, [] = _diagnostics} ->
+          groups
+          |> extract_targets(output_path, opts[:force])
+          |> generate(manifest(), output_path, opts)
+
+        {_groups, diagnostics} ->
+          {:error, diagnostics}
+      end
+
+    Enum.each(diagnostics, &print_diagnostic/1)
```

Comment:
Would it be nicer user experience if we could print errors as they happen? I think this makes the abstraction in the code less nice. Perhaps we could revisit that later.

Reply:
I figured there were three options:

1. **Print the errors as they happen in the Parser / Generator code.** This has the dual disadvantage of decentralizing error output and emitting output from functions outside of the mix context.
2. **Print errors as we receive them in the mix task.** I think this is what you're suggesting, and I think it would work fine but at the cost of spreading out the code a bit. We could still use `Diagnostic` as the normalized input structure into the error-printing code by converting all errors/warnings to `Diagnostic` first.
3. **Print errors as the end of the stream.** This felt like the simplest implementation because we're guaranteed that all diagnostics have been collected at this point. That's the approach you're seeing here.

I think (2) has some advantages because the error output is reported sooner, but I also felt like most error cases abort the compilation process early anyway, so in practice there might not be much of a delay between an error occurring and being reported.

Reply:
I think most cases aborting compilation is a solid argument for going with option (3).